### ISSUE TITLE
Add replication support

### ISSUE THREAD

**Issue by tinchodias** [pharo-nosql/voyage#28]
Voyage can take advantage of replication, an important feature in mongodb: https://docs.mongodb.org/manual/core/replication-introduction/

I want to refine this issue in more concrete terms. For the moment, I just paste a small replica-set demo: The following are commands to create a test replica set and test it (generated from https://docs.mongodb.org/manual/tutorial/deploy-replica-set-for-testing/)

	# terminal 1
	mongod --port 27030 --dbpath some_path/testreplicationdb/rs0-0 --replSet rs0 --smallfiles --oplogSize 128

	# terminal 2
	mongod --port 27031 --dbpath some_path/testreplicationdb/rs0-1 --replSet rs0 --smallfiles --oplogSize 128

	# terminal 3
	mongod --port 27032 --dbpath some_path/testreplicationdb/rs0-2 --replSet rs0 --smallfiles --oplogSize 128

	# terminal 4

	# log in to primary
	mongo --port 27030 
	rs.initiate() # make primary
	rs.conf() #just to check
	rs.add("my_host:27031")
	rs.add("my_host:27032")
	rs.status()  #just to check
	# add some data to primary
	db.categories.insert( { _id: "MongoDB", parent: "Databases" } )
	db.categories.insert( { _id: "Languages", parent: "Programming" } )

	#log in to a secondary
	mongo --port 27031 
	rs.slaveOk() #make readable
	db.categories.find() # gets:
	{ "_id" : "MongoDB", "parent" : "Databases" }
	{ "_id" : "Languages", "parent" : "Programming" }

**Comment by tinchodias** [comment]
I added support to read [config](https://docs.mongodb.com/manual/reference/command/replSetGetConfig/) and [status](https://docs.mongodb.com/manual/reference/command/replSetGetStatus/) in MongoTalk. It's unfinished yet, but you can check `MongoReplicationTest` class comment and hierarchy if you want.

**Comment by tinchodias** [comment]
This is a small demo that 
- connects to a secondary node
- reads it's replication set
- discovers members

```
mongo :=
	Mongo 
		host: 'localhost' 
		port: 27032.
mongo open.

replicaSet := mongo replicaSet.
 "a MongoReplicaSet(demo)"
replicaSet members.
 "an Array(a MongoReplicaSetMember(localhost:27031) a MongoReplicaSetMember(localhost:27032) a MongoReplicaSetMember(localhost:27033))"
replicaSet members collect: #priority. 
 "#(1.0 1.0 1.0)"

replicaSetStatus := mongo replicaSetStatus.
 "a MongoReplicaSetStatus(demo)"
replicaSetStatus members.
 "an Array(a MongoReplicaSetStatusMember(localhost:27031) a MongoReplicaSetStatusMember(localhost:27032) a MongoReplicaSetStatusMember(localhost:27033))"
replicaSetStatus members collect: #isPrimary. 
 "#(true false false)"
replicaSetStatus members collect: #stateString. 
 "#('PRIMARY' 'SECONDARY' 'SECONDARY')"

mongo close.
```

Both config (`replicaSet`) and status (`replicaSetStatus`) have relevant data. For example, config tells us the priority of the nodes, and status tells which node is actually the primary one (by default, the only node that accepts read/write persistence operations).
Note: I created the classes MongoReplicaSet and MongoReplicaSetStatus because config and status have different information... however, it might be a good idea to unify both (MongoReplicaSetMember and MongoReplicaSetStatusMember are very similar too)

**Comment by tinchodias** [comment]
I also [worked in a Voyage-level prototype](https://github.com/tinchodias/voyage/tree/replicaSetSupport) that is based on the "low-level" MongoTalk operations presented above.

My goal is to have a kind of VORepository that is aware of the replica-set and then performs the standard Voyage persistence operations to the "right" node. For example, if it was using a node that is now disconnected, it can automatically adapt to start using the node that replaced the former one.

In the cited branch, I created VOMongoReplicaSet, a subclass of VORepository that internally instantiates VOMongoRepository's to forward the persistence operations. However, now I think it is a better idea to take a slightly different approach: to implement the persistence-operation forwarding at the level of VOMongoRepositoryResolver. I mean, this new approach will use VOMongoRepository almost untouched but a new VOMongoRepositoryReplicaSetResolver. This way, the object cache as well as the MongoRepository options (versionGenerator, isEager, a future write strategy, etc.) are shared. I'm on this now, so this is work in progress but I decided to share it with you.

**Comment by tinchodias** [comment]
BTW, these are the steps to set up the demo mongo replication set. This might help to understand the small MongoTalk demo. (Note these steps are already commit together with mongo talk and voyage prototypes).

```
# set up some variables
baseRepo=/home/username/dev/voyage/testreplicationdb/
replSet=demo

# only first time: create db dirs
for i in `seq 1 3`;
do
	mkdir $baseRepo$replSet$i
done


# serve repos
for i in `seq 1 3`;
do
	xfce4-terminal --tab -T $replSet$i -e "mongod --port 2703$i --dbpath $baseRepo$replSet$i --replSet $replSet --smallfiles --oplogSize 128"
done

# wait them a bit
sleep 5s

# configure replication set
mongo --port 27031 --eval 'rs.initiate({ "_id" : '\"${replSet}\"', "members" : [ { "_id" : 1, "host" : "localhost:27031", } ], })'
# add secondaries
mongo --port 27031 --eval 'rs.add("localhost:27032")'
mongo --port 27031 --eval 'rs.add("localhost:27033")'
```

**Comment by tinchodias** [comment]
Good news! A first version of mongodb replication support in Voyage is available in [this commit](https://github.com/tinchodias/voyage/commit/dc4a43692445fd1c6246808f282925540ac5883d). 

It already has:
- detect if replica set is available on VOMongoRepository initialization
- if available, use the replica set for voyage operations (save, remove, selectAll,  etc.)
- automatic testing of standard voyage operations

It needs:
- manual testing (manually testing concrete scenarios) 
- support reading operations from other than the primary repository.
- ...

Starting point to test it: VOMongoReplicaSetTest; to understand the code: VOMongoReplicaSetRepositoryResolver.

**Comment by tinchodias** [comment]
Good news: with [my latest commit](https://github.com/tinchodias/voyage/commit/2d9d32d703d306bae2f8cce517937d647e709e1a), it is finally working in this main scenario:

Assume the replica set has 3 nodes (n1, n2, n3), where n1 is the primary (i.e. then the only one that supports read and write operations).

You can create now a VOMongoRepository on n3 and it will automatically resolve the operations to n1 (because it's the primary).

If n1 gets disconnected and mongodb elects n2 as the new primary, the same instance of  VOMongoRepository (on n3) will catch the errors on n1, then detect who is the new primary and finally start resolving the operations on n2.

All VORepositoryTest are green, and I have bash scripts to reproduce the scenarios in mongodb.

---

However, now I'm not sure if my design modifications are ok:
 
a) Now VOMongoRepositoryResolver is part of a hierarchy:

VOMongoRepositoryAbstractResolver
-- VOMongoRepositoryResolver
-- VOMongoReplicaSetRepositoryResolver

and the VOMongoRepository queries the mongodb to decide which one will be the resolver. This decision is done during initialization.

b) The VOMongoReplicaSetRepositoryResolver instance has a collection of instances a new subclass of VORepository called VOMongoReplicaSetRepository. This new repository is used internally by the VOMongoReplicaSetRepositoryResolver to forward the persistence operations.

Well, I do not think this explanation is clear enough so I would like to review and discuss it. 

---

I think main pending task is the implementation of some basic [read preference](https://docs.mongodb.com/manual/core/read-preference/) support... not much, only the capability to read from a secondary node.

**Comment by tinchodias** [comment]
Fix with #55
